### DISCUSSION TITLE
Combining `filterOr` and `mapOf`

### DISCUSSION THREAD

**Discussion by eroth** [TomWright/dasel#305] [Q&A] [answered]
Hi! Considering this response:

```json
[
    {
        "text": "Some great text",
        "reason": "44",
        "reactions": [
            {
                "name": "heart",
                "users": [
                    "U01JK7CB0TE"
                ],
                "count": 1
            },
            {
                "name": "pigeon",
                "users": [
                    "U01NMAR4LLE",
                    "U04KXRR5H6Z",
                    "U03DP251D9R"
                ],
                "count": 3
            },
            {
                "name": "car",
                "users": [
                    "U01NMAW4LLE",
                    "U04KXRD5H6Z",
                    "U03DP211D9R"
                ],
                "count": 3
            }
        ]
    }
]
```
I'd like to filter the `reactions` array so it only matches objects whose `name` field is `car` or `pigeon`, then pull out only the `text` field from the top-level object in the array, and output it all to JSON. So, output from this query would be:
```json
[
    {
        "text": "Some great text",
        "reactions": [
            {
                "name": "pigeon",
                "users": [
                    "U01NMAR4LLE",
                    "U04KXRR5H6Z",
                    "U03DP251D9R"
                ],
                "count": 3
            },
            {
                "name": "car",
                "users": [
                    "U01NMAW4LLE",
                    "U04KXRD5H6Z",
                    "U03DP211D9R"
                ],
                "count": 3
            }
        ]
    }
]
```
I can get the filtering part to work on the `reactions` array with
`dasel -r json 'all().reactions.all().filterOr(equal(name,car),equal(name,pigeon))'`

What I've tried (based on the [answer](https://github.com/TomWright/dasel/discussions/287#discussioncomment-4655686) to a previous question I've asked here) is:
`dasel -r json 'all().reactions.all().filterOr(equal(name,heart),equal(name,pigeon)).mapOf(text,parent(1).text)'`

My thoughts behind this were: Traverse to deepest node (`reactions` array), filter there, and then go up 1 level to the parent object which contains the `text` field and map that—but that doesn't seem to work.

Any suggestions on how I could do this? Thanks!

**Comment by TomWright** [comment, marked as answer]
I think this does what you expect:
```
$ cat test.json | dasel -r json 'all().reactions.all().filterOr(equal(name,car),equal(name,pigeon)).mapOf(text,parent(2).text,reactions,.).merge()'

[
  {
    "reactions": {
      "count": 3,
      "name": "pigeon",
      "users": [
        "U01NMAR4LLE",
        "U04KXRR5H6Z",
        "U03DP251D9R"
      ]
    },
    "text": "Some great text"
  },
  {
    "reactions": {
      "count": 3,
      "name": "car",
      "users": [
        "U01NMAW4LLE",
        "U04KXRD5H6Z",
        "U03DP211D9R"
      ]
    },
    "text": "Some great text"
  }
]
```

You just needed `parent(2)` instead of `parent(1)`.

`parent(1)` took the scope from here:
```
{
                "name": "pigeon",
                "users": [
                    "U01NMAR4LLE",
                    "U04KXRR5H6Z",
                    "U03DP251D9R"
                ],
                "count": 3
            }
```
to here:
```
[
            {
                "name": "pigeon",
                "users": [
                    "U01NMAR4LLE",
                    "U04KXRR5H6Z",
                    "U03DP251D9R"
                ],
                "count": 3
            },
            {
                "name": "car",
                "users": [
                    "U01NMAW4LLE",
                    "U04KXRD5H6Z",
                    "U03DP211D9R"
                ],
                "count": 3
            }
        ]
```

I would have expected some kind of error to have appeared, I'll have to look into that separately.

**Reply by eroth** [reply]
Ah, thank you so much! Yeah, I can see where I went wrong on the scope. With the `.` in `reactions,.`, is that some kind of placeholder for an array? I didn't see that anywhere in the docs. Thanks! Evan

**Reply by TomWright** [reply]
The `.reactions` accessed the reactions property, which contained an array.

**Reply by TomWright** [reply]
Sorry I misread the questions.

`reactions,.` in the context of `mapOf` says, create an item in the map with a key of `reactions`, and a value of `.` where `.` refers to the value of the current element in scope.

**Reply by TomWright** [reply]
The `.` usage is documented here https://daseldocs.tomwright.me/functions/this#usage
